### PR TITLE
add cmake support for Tpetra_ENABLE_DEPRECATED_CODE

### PR DESCRIPTION
**Pull request by GeoffDanielson**
This adds support to the CMakeLists.txt file to allow a macro switch to disable the old-style ordinal template args and dynamic allocation on matrices and graphs

@trilinos/tpetra 

## Description
Added a default value to the tpetra CMakeLists.txt file, support for the switch to the tpetra/src/core CMakeLists.txt, and a cmakedefine to TpetraCore_config.h.in

## Motivation and Context
This is an effort toward removing the template arguments (local/global/index ordinal types) and dynamic profile support from tpetra  without breaking backward compatibility with current apps

## How Has This Been Tested?
This was tested by running CMake with -DTpetra_ENABLE_DEPRECATED_CODE explicitly included in both (ON|OFF) states, and excluded to make sure that CMake built appropriately.

## Checklist

- [NA] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [NA] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [NA] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by kddevin** [comment]
BTW:  If you want the learning experience of adding new tests to Tpetra, you could write a simple one that CMake enables only if the new CMake option is enabled and does something like
```
main() {
#ifdef TPETRA_DISABLE_DEPRECATED_CODE
   printf("PASS");
#else 
   printf("FAIL");
#endif 
}
```
For when the CMake option isn't set, you could have a test with the opposite behavior.

File: packages/tpetra/CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,21 @@ TRIBITS_ADD_SHOW_DEPRECATED_WARNINGS_OPTION()
...
 
+# gcd 26 Oct 2018: Part of the great de-templating effort, 
+# allows users to switch at CMake time between using the old template
+# parameters, using Tpetra_DISABLE_DEPRECATED_CODE.  By default,
+# this will be set to "ON", to keep old behaviors.
+
+SET (Tpetra_DISABLE_DEPRECATED_CODE_DEFAULT OFF)
+MESSAGE (STATUS "Determine whether Tpetra will enable ordinal template arguments.")
+
+IF (Tpetra_DISABLE_DEPRECATED_CODE)
```
Comment (mhoemmen) [inline packages/tpetra/CMakeLists.txt]: Are you trying to define this option?  That's not how you do it.  Check out the definition of `Tpetra_ENABLE_CUDA` in `Trilinos/packages/tpetra/CMakeLists.txt` for an example.  I'll show you here how I would define `Tpetra_DISABLE_DEPRECATED_CODE`:
```
SET (Tpetra_DISABLE_DEPRECATED_CODE_DEFAULT OFF)

ASSERT_DEFINED (Tpetra_DISABLE_DEPRECATED_CODE_DEFAULT)
TRIBITS_ADD_OPTION_AND_DEFINE(
  Tpetra_DISABLE_DEPRECATED_CODE
  TPETRA_DISABLE_DEPRECATED_CODE
  "If ON, then disable deprecated code in Trilinos, at compile time."
  ${Tpetra_DISABLE_DEPRECATED_CODE_DEFAULT})

ASSERT_DEFINED (Tpetra_DISABLE_DEPRECATED_CODE)
IF (Tpetra_DISABLE_DEPRECATED_CODE)
  MESSAGE (STATUS "  - Tpetra_DISABLE_DEPRECATED_CODE is ON, so Tpetra will disable deprecated code at compile time.")
ELSE ()
  MESSAGE (STATUS "  - Tpetra_DISABLE_DEPRECATED_CODE is OFF, so Tpetra will not disable deprecated code at compile time.")
ENDIF ()
```

Comment (mhoemmen) [reply packages/tpetra/CMakeLists.txt]: Furthermore, this option is for all deprecated code, not just for deprecated template parameters.

Comment (mhoemmen) [reply packages/tpetra/CMakeLists.txt]: Remember that CMake has trinary logic:  `ON`, `OFF`, and `NOT DEFINED X`.  The latter doesn't quite mean the same thing as `OFF`.  This is why we take care to define the option before using it.

**Comment by mhoemmen** [comment]
I just marked this as a WIP, mainly so that it doesn't clog the testing pipeline, since we know what changes it needs.

File: packages/tpetra/CMakeLists.txt (hunk 1 above)
Comment (GeoffDanielson) [reply packages/tpetra/CMakeLists.txt]: Thanks, I added the support using TRIBITS_ADD_OPTION.

File: packages/tpetra/CMakeLists.txt [hunk 2]
```diff
@@ -15,18 +15,11 @@ TRIBITS_SET_ST_FOR_DEV_MODE(ST_FOR_DEV_PT_FOR_RELEASE)
...
-MESSAGE (STATUS "Determine whether Tpetra will enable ordinal template arguments.")
-
-IF (Tpetra_DISABLE_DEPRECATED_CODE)
-  MESSAGE (STATUS "  - Tpetra_DISABLE_DEPRECATED_CODE is ON, so Tpetra will assume you do not want the old template behavior.")
-  SET (Tpetra_DISABLE_DEPRECATED_CODE_DEFAULT ON)   
-ELSE()
-  MESSAGE (STATUS "  - Tpetra_DISABLE_DEPRECATED_CODE is OFF, so Tpetra will assume you want the old template behavior.")
-ENDIF()
+SET (Tpetra_ENABLE_DEPRECATED_CODE_DEFAULT ON)
+MESSAGE (STATUS "Determine whether Tpetra will enable ordinal template arguments and dynamic allocation.  By default, this will be ON.")
```
Comment (mhoemmen) [inline packages/tpetra/CMakeLists.txt]: Please just specify a general message here.  "Whether Tpetra enables deprecated code at compile time.  Default is ON (deprecated code enabled)."  We plan on using this macro in the future, for deprecating other things that we inevitably will want to deprecate :-) .  This also imitates Kokkos, which uses its macro in a generic way, for all deprecated features.

File: packages/tpetra/core/CMakeLists.txt [hunk 3]
```diff
@@ -14,6 +14,14 @@ TRIBITS_ADD_EXPLICIT_INSTANTIATION_OPTION()
 # template instantiation (ETI).
 TRIBITS_ADD_ETI_SUPPORT()
 
+# adding macro for deprecated code (templated L0 and G0, and dynamic allocation)
+TRIBITS_ADD_OPTION_AND_DEFINE(
+  Tpetra_ENABLE_DEPRECATED_CODE
+  TPETRA_ENABLE_DEPRECATED_CODE
+  "Disable Tpetra deprecated code (templated ordinal types and dynamic graph/matrix allocation)"
+  Tpetra_ENABLE_DEPRECATED_CODE_DEFAULT
```
Comment (mhoemmen) [inline packages/tpetra/core/CMakeLists.txt]: I think you need to dereference the default value here.  Do this:
```
${Tpetra_ENABLE_DEPRECATED_CODE_DEFAULT}
```

File: packages/tpetra/core/cmake/TpetraCore_config.h.in [hunk 4]
```diff
@@ -2,6 +2,9 @@
 #define TPETRACORE_CONFIG_H
 /* CMake uses this file to generate TpetraCore_config.h automatically */
 
+/* define if user is using the deprecated codebase (templated ordinals, dynamicprofile) or not */
```
Comment (mhoemmen) [inline packages/tpetra/core/cmake/TpetraCore_config.h.in]: See above.  "Deprecated code" is a generic thing; we should not refer to specifically deprecated features here.

File: packages/tpetra/CMakeLists.txt (hunk 2 above)
Comment (GeoffDanielson) [reply packages/tpetra/CMakeLists.txt]: Got it.  Thanks

File: packages/tpetra/core/CMakeLists.txt (hunk 3 above)
Comment (GeoffDanielson) [reply packages/tpetra/core/CMakeLists.txt]: You're right, I did need that for the default.

File: packages/tpetra/core/cmake/TpetraCore_config.h.in (hunk 4 above)
Comment (GeoffDanielson) [reply packages/tpetra/core/cmake/TpetraCore_config.h.in]: Changed.

**Comment by mhoemmen** [comment]
@kddevin It looks like Geoff addressed your concerns; shall we go ahead with the merge?

**Pull request merged**